src/Diagnostics/FxCop/Desktop.Analyzers/Core/Security/DoNotCatchCorruptedStateExceptionsAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+                                    codeBlockStartContext.RegisterCodeBlockEndAction(analyzer.AnalyzeCodeBlockEnd);
+                                }
+                            }
+                        }); 
+                    }
+                });
+        }
+
+        protected abstract class Analyzer
+        {
+            private readonly ISymbol _owningSymbol;
+            private readonly SyntaxNode _codeBlock;
+            private readonly Dictionary<TCatchClauseSyntax, ISymbol> _catchAllCatchClauses;
+
+            public abstract TLanguageKindEnum CatchClauseKind { get; }
+            public abstract TLanguageKindEnum ThrowStatementKind { get; }
+            protected CompilationSecurityTypes TypesOfInterest { get; private set; }
+            protected abstract ISymbol GetExceptionTypeSymbolFromCatchClause(TCatchClauseSyntax catchNode, SemanticModel model);
+            protected abstract bool IsThrowStatementWithNoArgument(TThrowStatementSyntax throwNode);
+            protected abstract bool IsCatchClause(SyntaxNode node);
+            protected abstract bool IslambdaExpression(SyntaxNode node);
+
+            protected Analyzer(CompilationSecurityTypes compilationTypes, ISymbol owningSymbol, SyntaxNode codeBlock)
+            {
+                _owningSymbol = owningSymbol;
+                _codeBlock = codeBlock;
+                _catchAllCatchClauses = new Dictionary<TCatchClauseSyntax, ISymbol>();
+                TypesOfInterest = compilationTypes;
+            }
+
+            public void AnalyzeCatchClause(SyntaxNodeAnalysisContext context)
+            {
+                var catchNode = (TCatchClauseSyntax)context.Node;
+                ISymbol exceptionTypeSymbol = GetExceptionTypeSymbolFromCatchClause(catchNode, context.SemanticModel);
+
+                if (IsCatchTypeTooGeneral(exceptionTypeSymbol))
+                {
+                    var parentNode = catchNode.Parent;
+                    while (parentNode != _codeBlock)
+                    {
+                        // for now there doesn't seem to have any way to annotate lambdas with attributes
+                        if (IslambdaExpression(parentNode))
+                        {
+                            return;
+                        }
+                        parentNode = parentNode.Parent;
+                    }
+                    _catchAllCatchClauses[catchNode] = exceptionTypeSymbol;
+                }
+            }
+
+            public void AnalyzeThrowStatement(SyntaxNodeAnalysisContext context)
+            {
+                var throwNode = (TThrowStatementSyntax)context.Node;
+
+                // throwNode is a throw statement with no argument, which is not allowed outside of a catch clause
+                if (IsThrowStatementWithNoArgument(throwNode))
+                {
+                    var enlcosingCatchClause = (TCatchClauseSyntax)throwNode.Ancestors().First(IsCatchClause);
```

Comment:
@mavasani So I assume we should avoid ancestors/descendants traversal call *only if it is possible*? For example, the logic here is to find the enclosing catch clause for the throw statement, since there's no callback at the end of syntax node analysis, I don't see a way to rewrite this logic without uplink traversal.